dkist/dataset/loader.py

```diff
@@ -0,0 +1,75 @@
+import importlib.resources as importlib_resources
+from pathlib import Path
+from functools import singledispatch
+
+from jsonschema.exceptions import ValidationError
+
+import asdf
+
+
+@singledispatch
+def load_dataset(target):
+    """Function to load a Dataset from either an asdf file path or a directory."""
+    raise TypeError("""Input type not recognised. Must be a string or pathlib.Path referencing a
+        .asdf file or a directory containing one.""")
```

Comment:
This indentation makes my head hurt.
```suggestion
    raise TypeError("Input type not recognised. Must be a string or pathlib.Path referencing a
                    .asdf file or a directory containing one.")
```